Python/bytecodes.c

```diff
@@ -2222,16 +2227,36 @@ dummy_func(
             PyObject *attr_o;
 
             PyDictObject *dict = _PyObject_GetManagedDict(owner_o);
-            DEOPT_IF(hint >= (size_t)dict->ma_keys->dk_nentries);
+            DEOPT_IF(!LOCK_OBJECT(dict));
+            #ifdef Py_GIL_DISABLED
+            if (dict != _PyObject_GetManagedDict(owner_o)) {
+                UNLOCK_OBJECT(dict);
+                DEOPT_IF(true);
+            }
+            #endif
+            if (hint >= (size_t)dict->ma_keys->dk_nentries) {
+                UNLOCK_OBJECT(dict);
+                DEOPT_IF(true);
+            }
             PyObject *name = GETITEM(FRAME_CO_NAMES, oparg>>1);
-            DEOPT_IF(!DK_IS_UNICODE(dict->ma_keys));
+            if (!DK_IS_UNICODE(dict->ma_keys)) {
```

Comment:
I think this should be an explicit check that the kind is `DICT_KEYS_UNICODE`. Currently, this allows for split keys, which will deopt later when the `me_value` is `NULL`. Excluding split keys early makes the thread-safety analysis easier (and isn't any more expensive).